File: PInvoke/Controls/TranslateSnippetControl.vb
' Copyright (c) Microsoft Corporation.  All rights reserved.
Imports System.Collections.ObjectModel
Imports System.ComponentModel
Imports System.Windows.Forms
Imports PInvoke.Parser
Imports PInvoke.Transform

Namespace Controls

    Public Class TranslateSnippetControl

        Private Class RequestData
            Friend Text As String
            Friend InitialMacroList As ReadOnlyCollection(Of Macro)
        End Class

        Private Class ResponseData
            Friend ParseOutput As String
        End Class

        Private _ns As NativeStorage = NativeStorage.DefaultInstance
        Private _transKind As TransformKindFlags = TransformKindFlags.All
        Private _initialMacroList As New List(Of Macro)
        Private _changed As Boolean

        Public Property AutoGenerate() As Boolean
            Get
                Return m_autoGenerateBtn.Checked
            End Get
            Set(ByVal value As Boolean)
                m_autoGenerateBtn.Checked = value
            End Set
        End Property

        Public Sub New()

            ' This call is required by the Windows Form Designer.
            InitializeComponent()

            ' Add any initialization after the InitializeComponent() call.
            m_langTypeCb.Items.AddRange(PInvoke.EnumUtil.GetAllValuesObject(Of LanguageType))
            m_langTypeCb.SelectedItem = LanguageType.VisualBasic
        End Sub

#Region "ISignatureImportControl"

        ''' <summary>
        ''' Language that we are displaying the generated values in
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property LanguageType() As LanguageType Implements ISignatureImportControl.LanguageType
            Get
                If m_langTypeCb.SelectedItem Is Nothing Then
                    Return Transform.LanguageType.VisualBasic
                End If

                Return DirectCast(m_langTypeCb.SelectedItem, LanguageType)
            End Get
            Set(ByVal value As LanguageType)
                m_langTypeCb.SelectedItem = value
            End Set
        End Property

        <DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)>
        Public Property NativeStorage() As NativeStorage Implements ISignatureImportControl.NativeStorage
            Get
                Return _ns
            End Get
            Set(ByVal value As NativeStorage)
                _ns = value
                _initialMacroList = Nothing
            End Set
        End Property

        <DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)>
        Public Property TransformKindFlags() As TransformKindFlags Implements ISignatureImportControl.TransformKindFlags
            Get
                Return _transKind
            End Get
            Set(ByVal value As TransformKindFlags)
                _transKind = value
            End Set
        End Property

        Public Event LanguageTypeChanged As EventHandler Implements ISignatureImportControl.LanguageTypeChanged

        Public ReadOnly Property ManagedCode() As String Implements ISignatureImportControl.ManagedCode
            Get
                Return m_managedCodeBox.Text
            End Get
        End Property

#End Region

#Region "Event Handlers"

        Private Sub OnNativeCodeChanged(ByVal sender As Object, ByVal e As EventArgs) Handles m_nativeCodeTb.TextChanged
            If m_bgWorker.IsBusy Then
                _changed = True
            Else
                RunWorker()
            End If
        End Sub

        Private Shared Sub OnDoBackgroundWork(ByVal sender As System.Object, ByVal e As System.ComponentModel.DoWorkEventArgs) Handles m_bgWorker.DoWork
            Dim result As New ResponseData
            Try
                Dim req As RequestData = DirectCast(e.Argument, RequestData)
                Dim code As String = req.Text
                Dim analyzer As NativeCodeAnalyzer = NativeCodeAnalyzerFactory.CreateForMiniParse(OsVersion.WindowsVista, req.InitialMacroList)
                Using reader As New IO.StringReader(code)
                    Dim parseResult As NativeCodeAnalyzerResult = analyzer.Analyze(reader)
                    Dim ep As ErrorProvider = parseResult.ErrorProvider
                    If ep.Warnings.Count = 0 AndAlso ep.Errors.Count = 0 Then
                        result.ParseOutput = "None ..."
                    Else
                        result.ParseOutput = ep.CreateDisplayString()
                    End If
                End Using
            Catch ex As Exception
                result.ParseOutput = ex.Message
            End Try

            e.Result = result
        End Sub

        Private Sub OnBackgroundOperationCompleted(ByVal sender As System.Object, ByVal e As System.ComponentModel.RunWorkerCompletedEventArgs) Handles m_bgWorker.RunWorkerCompleted
            Dim response As ResponseData = DirectCast(e.Result, ResponseData)
            m_errorsTb.Text = response.ParseOutput
            If _changed Then
                RunWorker()
            ElseIf m_autoGenerateBtn.Checked Then
                GenerateCode()
            End If

        End Sub

        Private Sub OnGenerateCodeClick(ByVal sender As Object, ByVal e As EventArgs) Handles m_generateBtn.Click
            GenerateCode()
        End Sub

        Private Sub OnAutoGenerateCodeCheckChanged(ByVal sender As Object, ByVal e As EventArgs) Handles m_autoGenerateBtn.CheckedChanged
            If m_autoGenerateBtn.Checked Then
                GenerateCode()
            End If
        End Sub

        Private Sub OnLanguageTypeChanged(ByVal sender As Object, ByVal e As EventArgs) Handles m_langTypeCb.SelectedIndexChanged
            If m_autoGenerateBtn.Checked Then
                GenerateCode()
            End If

            RaiseEvent LanguageTypeChanged(Me, EventArgs.Empty)
        End Sub

        Private Sub m_nativeCodeTb_KeyDown(ByVal sender As System.Object, ByVal e As System.Windows.Forms.KeyEventArgs) Handles m_nativeCodeTb.KeyDown
            If e.KeyCode = Keys.A And e.Modifiers = Keys.Control Then
                m_nativeCodeTb.SelectAll()
                e.Handled = True
            End If
        End Sub

#End Region

#Region "Helpers"

        Private Sub RunWorker()
            _changed = False
            m_errorsTb.Text = "Parsing ..."
            If _initialMacroList Is Nothing Then
                _initialMacroList = _ns.LoadAllMacros()
            End If

            Dim data As New RequestData()
            data.InitialMacroList = New ReadOnlyCollection(Of Macro)(_initialMacroList)
            data.Text = m_nativeCodeTb.Text
            m_bgWorker.RunWorkerAsync(data)
        End Sub

        Private Sub GenerateCode()
            Try
                Dim conv As New BasicConverter(LanguageType, _ns)
                conv.TransformKindFlags = _transKind
                m_managedCodeBox.Code = conv.ConvertNativeCodeToPInvokeCode(m_nativeCodeTb.Text)
            Catch ex As Exception
                m_managedCodeBox.Code = ex.Message
            End Try
        End Sub

#End Region

    End Class

End Namespace

File: PInvoke.Core/Controls/TranslateSnippetControl.cs
// Copyright (c) Microsoft Corporation.  All rights reserved.

using System;
using System.Linq;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Windows.Forms;
using PInvoke.Parser;
using PInvoke.Transform;
using System.IO;

namespace PInvoke.Controls
{
    public partial class TranslateSnippetControl
    {
        private class RequestData
        {
            internal string Text;
            internal ReadOnlyCollection<Macro> InitialMacroList;
        }

        private class ResponseData
        {
            internal string ParseOutput;
        }

        private INativeSymbolStorage _storage;
        private TransformKindFlags _transKind = TransformKindFlags.All;
        private List<Macro> _initialMacroList = new List<Macro>();

        private bool _changed;
        public bool AutoGenerate
        {
            get { return m_autoGenerateBtn.Checked; }
            set { m_autoGenerateBtn.Checked = value; }
        }


        public TranslateSnippetControl()
        {
            // This call is required by the Windows Form Designer.
            InitializeComponent();

            _storage = new BasicSymbolStorage();

            // Add any initialization after the InitializeComponent() call.
            m_langTypeCb.Items.AddRange(EnumUtil.GetAllValuesObject<LanguageType>());
            m_langTypeCb.SelectedItem = LanguageType.VisualBasic;
        }

        public TranslateSnippetControl(INativeSymbolStorage storage)
        {
            _storage = storage;
        }

        #region "ISignatureImportControl"

        /// <summary>
        /// Language that we are displaying the generated values in
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public LanguageType LanguageType
        {
            get
            {
                if (m_langTypeCb.SelectedItem == null)
                {
                    return Transform.LanguageType.VisualBasic;
                }

                return (LanguageType)m_langTypeCb.SelectedItem;
            }
            set { m_langTypeCb.SelectedItem = value; }
        }

        public INativeSymbolStorage Storage
        {
            get { return _storage; }
            set
            {
                _storage = value;
                _initialMacroList = null;
            }
        }

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
        public TransformKindFlags TransformKindFlags
        {
            get { return _transKind; }
            set { _transKind = value; }
        }

        public event EventHandler LanguageTypeChanged;

        public string ManagedCode
        {
            get { return m_managedCodeBox.Text; }
        }

        #endregion

        #region "Event Handlers"

        private void OnNativeCodeChanged(object sender, EventArgs e)
        {
            if (m_bgWorker.IsBusy)
            {
                _changed = true;
            }
            else
            {
                RunWorker();
            }
        }

        private static void OnDoBackgroundWork(System.Object sender, System.ComponentModel.DoWorkEventArgs e)
        {
            ResponseData result = new ResponseData();
            try
            {
                RequestData req = (RequestData)e.Argument;
                string code = req.Text;
                NativeCodeAnalyzer analyzer = NativeCodeAnalyzerFactory.CreateForMiniParse(OsVersion.WindowsVista, req.InitialMacroList);
                analyzer.IncludePathList.Add("c:\\program files (x86)\\windows kits\\8.1\\include\\shared");
                using (var reader = new StringReader(code))
                {
                    NativeCodeAnalyzerResult parseResult = analyzer.Analyze(reader);
                    ErrorProvider ep = parseResult.ErrorProvider;
                    if (ep.Warnings.Count == 0 && ep.Errors.Count == 0)
                    {
                        result.ParseOutput = "None ...";
                    }
                    else
                    {
                        result.ParseOutput = ep.CreateDisplayString();
                    }
                }
            }
            catch (Exception ex)
            {
                result.ParseOutput = ex.Message;
            }

            e.Result = result;
        }

        private void OnBackgroundOperationCompleted(System.Object sender, System.ComponentModel.RunWorkerCompletedEventArgs e)
        {
            ResponseData response = (ResponseData)e.Result;
            m_errorsTb.Text = response.ParseOutput;
            if (_changed)
            {
                RunWorker();
            }
            else if (m_autoGenerateBtn.Checked)
            {
                GenerateCode();
            }

        }

        private void OnGenerateCodeClick(object sender, EventArgs e)
        {
            GenerateCode();
        }

        private void OnAutoGenerateCodeCheckChanged(object sender, EventArgs e)
        {
            if (m_autoGenerateBtn.Checked)
            {
                GenerateCode();
            }
        }

        private void OnLanguageTypeChanged(object sender, EventArgs e)
        {
            if (m_autoGenerateBtn.Checked)
            {
                GenerateCode();
            }

            if (LanguageTypeChanged != null)
            {
                LanguageTypeChanged(this, EventArgs.Empty);
            }
        }

        private void m_nativeCodeTb_KeyDown(System.Object sender, System.Windows.Forms.KeyEventArgs e)
        {
            if (e.KeyCode == Keys.A & e.Modifiers == Keys.Control)
            {
                m_nativeCodeTb.SelectAll();
                e.Handled = true;
            }
        }

        #endregion

        #region "Helpers"

        private void RunWorker()
        {
            _changed = false;
            m_errorsTb.Text = "Parsing ...";
            if (_initialMacroList == null)
            {
                _initialMacroList = _storage.GetAllMacros().ToList();
            }

            RequestData data = new RequestData();
            data.InitialMacroList = new ReadOnlyCollection<Macro>(_initialMacroList);
            data.Text = m_nativeCodeTb.Text;
            m_bgWorker.RunWorkerAsync(data);
        }

        private void GenerateCode()
        {
            try
            {
                var conv = new BasicConverter(LanguageType, _storage);
                conv.TransformKindFlags = _transKind;
                m_managedCodeBox.Code = conv.ConvertNativeCodeToPInvokeCode(m_nativeCodeTb.Text);
            }
            catch (Exception ex)
            {
                m_managedCodeBox.Code = ex.Message;
            }
        }

        #endregion

    }

}

//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================
